packages/aws-cdk/lib/api/hotswap/lambda-functions.ts

```diff
@@ -0,0 +1,159 @@
+import * as cfn_diff from '@aws-cdk/cloudformation-diff';
+import { ISDK } from '../aws-auth';
+import { assetMetadataChanged, ChangeHotswapImpact, ChangeHotswapResult, HotswapOperation, ListStackResources, stringifyPotentialCfnExpression } from './common';
+
+/**
+ * Returns `false` if the change cannot be short-circuited,
+ * `true` if the change is irrelevant from a short-circuit perspective
+ * (like a change to CDKMetadata),
+ * or a LambdaFunctionResource if the change can be short-circuited.
+ */
+export function isHotswappableLambdaFunctionChange(
+  logicalId: string, change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): ChangeHotswapResult {
+  const lambdaCodeChange = isLambdaFunctionCodeOnlyChange(change, assetParamsWithEnv);
+  if (typeof lambdaCodeChange === 'string') {
+    return lambdaCodeChange;
+  } else {
+    // verify that the Asset changed - otherwise,
+    // it's a Code property-only change,
+    // but not to an asset change
+    // (for example, going from Code.fromAsset() to Code.fromInline())
+    if (!assetMetadataChanged(change)) {
+      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    }
+
+    let functionPhysicalName: string | undefined;
+    try {
+      functionPhysicalName = stringifyPotentialCfnExpression(change.newValue?.Properties?.FunctionName, assetParamsWithEnv);
+    } catch (e) {
+      // It's possible we can't evaluate the function's name -
+      // for example, it can use a Ref to a different resource,
+      // which we wouldn't have in `assetParamsWithEnv`.
+      // That's fine though - ignore any errors,
+      // and treat this case the same way as if the name wasn't provided at all,
+      // which means it will be looked up using the listStackResources() call
+      // by the later phase (which actually does the Lambda function update)
+      functionPhysicalName = undefined;
+    }
+
+    return new LambdaFunctionHotswapOperation({
+      logicalId,
+      physicalName: functionPhysicalName,
+      code: lambdaCodeChange,
+    });
+  }
+}
+
+/**
+ * Returns `true` if the change is not for a AWS::Lambda::Function,
+ * but doesn't prevent short-circuiting
+ * (like a change to CDKMetadata resource),
+ * `false` if the change is to a AWS::Lambda::Function,
+ * but not only to its Code property,
+ * or a LambdaFunctionCode if the change is to a AWS::Lambda::Function,
+ * and only affects its Code property.
+ */
+function isLambdaFunctionCodeOnlyChange(
+  change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): LambdaFunctionCode | ChangeHotswapImpact {
+  if (!change.newValue) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+  const newResourceType = change.newValue.Type;
+  // Ignore Metadata changes
+  if (newResourceType === 'AWS::CDK::Metadata') {
```

Comment:
I don't understand why this is special-cased? What's different between a Lambda changing its type to an `AWS::S3::Bucket` vs changing its type to an `AWS::CDK::Metadata`? 

Also, this doesn't feel correct. If the Type actually changed to an `AWS::CDK::Metadata`, wouldn't the original Lambda function have to be destroyed? Which also requires a full deployment.

Reply:
This check is not about a resource changing its type from an `AWS::Lambda::Function` to a `AWS::CDK::Metadata` - this is just about checking its current type.

Reply:
Then why is it in `isLambdaFunctionCodeOnlyChange` ?

Reply:
Oh I see, the function is probably poorly named. This is where you would do extension points for other resource types, right?

Reply:
Yes.